### PR TITLE
[GEOT-6958]: imageMosaic fix on mixed coverages (heterogeneous+homogeneous)

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOT-6958](https://osgeo-org.atlassian.net/browse/GEOT-6958) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

See https://osgeo-org.atlassian.net/browse/GEOT-6958

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicReader.java [hunk 1]
```diff
@@ -385,7 +386,18 @@ private void initReaderFromURL(final Object source, final Hints hints) throws Ex
                             ImageMosaicConfigHandler.createGranuleCatalogFromDatastore(
                                     parentDirectory, datastoreProperties, true, getHints());
                 } else {
-                    CatalogConfigurationBean bean = beans.get(0).getCatalogConfigurationBean();
+                    // We need to figure out if there is an heterogeneous mosaic in the list and use
+                    // it as reference. (Otherwise, read on heterogeneous coverages may not work
+                    // properly)
+                    List<MosaicConfigurationBean> referenceBeans =
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicReader.java:392]: You actually just want the first no? Then it's slightly better to use findFirst and then check the returned Optional, and use the "orElse" to pick the first from the original list as a fallback.

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicReaderTest.java [hunk 2]
```diff
@@ -4632,6 +4634,40 @@ public void testCoverageOnBands() throws Exception {
         reader.dispose();
     }
 
+    @Test
+    public void testHeterogeneousConfigs() throws Exception {
+        // we have 2 mosaics here:
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicReaderTest.java:4639]: Two mosaics... or two coverages in the same mosaic?

**Review by aaime — CHANGES_REQUESTED** [review]
Makes sense, proposed a small improvement.

File: modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicReaderTest.java (hunk 2 above)
Comment (dromagnoli) [reply modules/plugin/imagemosaic/src/test/java/org/geotools/gce/imagemosaic/ImageMosaicReaderTest.java:4639]: Right, they are actually 2 coverages of the same mosaic store.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicReader.java (hunk 1 above)
Comment (dromagnoli) [reply modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicReader.java:392]: Oki, Will do.
Thanks for the feedback

**Comment by dromagnoli** [comment]
@aaime all feedbacks have been addressed.

**Review by aaime — APPROVED** [review]

**Pull request merged**